drivers/dac/dac_ad569x.c

```diff
@@ -0,0 +1,239 @@
+
+enum ad569x_command {
+	AD569X_CMD_WRITE = 0x10,
+	AD569X_CMD_UPDATE = 0x20,
+	AD569X_CMD_WRITE_AND_UPDATE = 0x30,
+	AD569X_CMD_CONFIGURE = 0x40,
+};
+
+struct ad569x_config {
+	struct i2c_dt_spec bus;
+	uint8_t resolution;
+	uint8_t gain;
+	uint8_t voltage_reference;
+	uint8_t power_down_mode;
+};
+
+enum {
+	AD569X_CTRL_GAIN_1 = 0,
+	AD569X_CTRL_GAIN_2 = 1,
+	AD569X_CTRL_GAIN_DEFAULT = AD569X_CTRL_GAIN_1,
+	AD569X_CTRL_GAIN_MAX = AD569X_CTRL_GAIN_2,
+};
+
+enum {
+	AD569X_CTRL_REF_INTERNAL = 0,
+	AD569X_CTRL_REF_EXTERNAL = 1,
+	AD569X_CTRL_REF_DEFAULT = AD569X_CTRL_REF_INTERNAL,
+	AD569X_CTRL_REF_MAX = AD569X_CTRL_REF_EXTERNAL,
+};
+
+enum {
+	AD569X_CTRL_PD_NORMAL = 0,
+	AD569X_CTRL_PD_1K_OUT_IMPEDANCE = 1,
+	AD569X_CTRL_PD_100K_OUT_IMPEDANCE = 2,
+	AD569X_CTRL_PD_3_STATE_OUT_IMPEDANCE = 3,
+	AD569X_CTRL_PD_DEFAULT = AD569X_CTRL_PD_NORMAL,
+	AD569X_CTRL_PD_MAX = AD569X_CTRL_PD_3_STATE_OUT_IMPEDANCE,
+};
+
+enum {
+	AD569X_CTRL_NO_RESET = 0,
+	AD569X_CTRL_PERFORM_RESET = 1,
+	AD569X_CTRL_RESET_DEFAULT = AD569X_CTRL_NO_RESET
+};
+
+static int ad569x_write(const struct device *dev, enum ad569x_command command, uint16_t value)
+{
+	const struct ad569x_config *config = dev->config;
+
+	uint8_t tx_data[3];
+
+	tx_data[0] = command;
+	sys_put_be16(value, tx_data + 1);
+
+	return i2c_write_dt(&config->bus, tx_data, sizeof(tx_data));
+}
+
+static int ad569x_read(const struct device *dev, uint16_t *value)
+{
+	const struct ad569x_config *config = dev->config;
+
+	uint8_t rx_data[2];
+	int result;
+
+	result = i2c_read_dt(&config->bus, rx_data, sizeof(rx_data));
+	if (result == 0) {
+		*value = sys_get_be16(rx_data);
+	}
+
+	return result;
+}
+
+static int ad569x_channel_setup(const struct device *dev, const struct dac_channel_cfg *channel_cfg)
+{
+	const struct ad569x_config *config = dev->config;
+
+	if (channel_cfg->channel_id > DAC_AD569X_MAX_CHANNEL) {
+		LOG_ERR("invalid channel %d", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->resolution != config->resolution) {
+		LOG_ERR("invalid resolution %d", channel_cfg->resolution);
+		return -EINVAL;
+	}
+
+	if (config->gain > AD569X_CTRL_GAIN_MAX) {
+		LOG_ERR("invalid gain %d", config->gain);
+		return -EINVAL;
+	}
+
+	if (config->voltage_reference > AD569X_CTRL_REF_MAX) {
+		LOG_ERR("invalid voltage reference %d", config->voltage_reference);
+		return -EINVAL;
+	}
+
+	if (config->power_down_mode > AD569X_CTRL_PD_MAX) {
+		LOG_ERR("invalid power down mode %d", config->power_down_mode);
+		return -EINVAL;
+	}
+
+	LOG_DBG("configure %s: gain %d, voltage reference %d, power down mode %d", dev->name,
+		config->gain, config->voltage_reference, config->power_down_mode);
+
+	uint16_t ctrl_reg = AD569X_CTRL_GAIN(config->gain) |
+			    AD569X_CTRL_REF(config->voltage_reference) |
+			    AD569X_CTRL_PD(config->power_down_mode);
+
+	int result = ad569x_write(dev, AD569X_CMD_CONFIGURE, ctrl_reg);
+
+	if (result != 0) {
+		LOG_ERR("failed to configure the device");
+		return result;
+	}
```

Comment:
The checks should be done at compile-time (ideally through the Devicetree binding as mentioned above), as the `config` is const and will never change during operation.

Writing the default configuration should be done in `ad569x_init`, as this is not channel-related.